pennylane/optimize/rotoselect.py

```diff
@@ -0,0 +1,178 @@
+# Copyright 2018-2020 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Rotoselect gradient free optimizer"""
+
+import pennylane as qml
+from pennylane import numpy as np
+
+
+class RotoselectOptimizer:
+    r"""Rotoselect gradient free optimizer.
+
+    The Rotoselect optimizer minimizes an objective function with respect to the parameters and
+    gates of a quantum circuit without the need of calculating the gradient of the function. The
+    algorithm works by updating the parameters :math:`\theta = \theta_1, \dots, \theta_D` and gate
+    choices :math:`R = R_1,\dots,R_D` one at a time according to a closed-form expression for the
+    optimal value of the :math:`d^{th}` parameter :math:`\theta^*_d` when the other parameters and
+    gate choices are fixed:
+
+    .. math:: \theta^*_d = \underset{\theta_d}{\text{argmin}}\left<H\right>_{\theta_d}
+              = -\frac{\pi}{2} - \text{arctan2}\left(2\left<H\right>_{\theta_d=0}
+              - \left<H\right>_{\theta_d=\pi/2} - \left<H\right>_{\theta_d=-\pi/2},
+              \left<H\right>_{\theta_d=\pi/2} - \left<H\right>_{\theta_d=-\pi/2}\right),
+
+    where :math:`\left<H\right>_{\theta_d}` is the expectation value of the objective function
+    optimized over the parameter :math:`\theta_d`. :math:`\text{arctan2}(x, y)` computes the
+    element-wise arc tangent of :math:`x/y` choosing the quadrant correctly, avoiding, in
+    particular, division-by-zero when :math:`y = 0`.
+
+    The algorithm is described in further detail in `Ostaszewski et al. (2019) <https://arxiv.org/abs/1905.09692>`_.
+
+    Keyword Args:
+        possible_generators (list[~.Operation]): List containing the possible ``pennylane.ops.qubit`` operators
+            that are allowed in the circuit. Default is the set of Pauli matrices :math:`\{X, Y, Z\}`.
+
+    **Example:**
+
+    Initialize the optimizer, set the initial values of ``x``, choose the initial generators to
+    start with, and set the number of steps to optimize over.
+
+    >>> opt = qml.optimize.RotoselectOptimizer()
+    >>> x = [0.3, 0.7]
+    >>> generators = [qml.RX, qml.RY]
+    >>> n_steps = 1000
+
+    Define a cost function based on a PennyLane circuit
```

Comment:
It would be good here to specify exactly how the generators fit into the PennyLane QNode, and note that the QNode _must_ satisfy the signature restriction. This could be as simple as a note box,

```rest
.. note::

    The QNode included in the optimization must have have a signature matching
    ``qnode(x, generators=None)``, where ``x`` are the trainable weights of the variational
    circuit, and `generators` is a list of parametrized PennyLane operators.
```

It would also be useful to include a very small QNode example before defining the cost function.

A couple of other questions I have that are maybe not so clear in the example:

* Does the number of generators in the list need to match the number of weights? i.e., what happens if some of the trainable parameters are using fixed gates?

* Does it only work with Pauli rotations (RX, RY, RZ)? How about controlled rotations (CRX, CRZ, etc.), or gates that accept multiple parameters (U1, Rot)?